scripts/exact_1d_bayesian_filtering_demo.py

```diff
@@ -0,0 +1,515 @@
+
+import jax
+from jax import numpy as jnp
+from jax import scipy as jsp
+from jax import random
+import matplotlib.pyplot as plt
+
+
+plt.rcParams.update({"font.size": 16})
+
+
+def generate_x_true(
+        rng_key: jnp.DeviceArray, max_iter: int, x0_rvs: Callable,
+        v_rvs: Callable, f: Callable):
+    rng_keys = random.split(rng_key, num=2)
+    x0 = x0_rvs(rng_keys[0], shape=())
+    x_true = jnp.zeros(shape=(max_iter + 1,))
+    x_true = x_true.at[0].set(x0)
+    v = v_rvs(rng_keys[1], shape=x_true.shape)
+    for k in range(1, max_iter + 1):
+        x_true = x_true.at[k].set(f(x_true[k - 1], v[k - 1], k=k - 1))
+    return jnp.array(x_true)
+
+
+def generate_y(
+        rng_key: jnp.DeviceArray, x_true: jnp.DeviceArray, e_rvs: Callable,
+        h: Callable):
+    shape = x_true.shape
+    e = e_rvs(rng_key, shape=shape)
+    y = h(x_true, e)
+    y = y.at[0].set(jnp.inf)
+    return y
+
+
+def x_pdf(x_new, x, k, v_pdf, f):
+    v = x_new - f(x=x, v=0, k=k)
+    return v_pdf(x - v)
+
+
+def y_likelihood(y, x, e_pdf, h):
+    e = h(x=x, e=0)
+    return e_pdf(y - e)
+
+
+def inversion_sampling(rng_key, x_grid, px_grid, num_samples):
+    rng_keys = random.split(rng_key, num=2)
+    u = random.uniform(rng_keys[0], shape=(num_samples, 1))
+    delta = x_grid[1] - x_grid[0]
+    noise = random.uniform(
+        rng_keys[1], minval=-delta / 2, maxval=delta / 2, shape=(num_samples,)
+    )
+    # It only works for sufficient dense uniformly spaced grid
+    point_mass = px_grid
+
+    cdf = jnp.cumsum(point_mass)
+
+    bound_cdf = jnp.where(cdf < u, cdf, 0)
+    idx = jnp.argmax(bound_cdf, axis=1)
+    x = x_grid[idx]
+    return x + noise
+
+
+def kde(x_grid, x, kernel_variance):
+    delta = x_grid[1] - x_grid[0]
+    # broadcast it into (n_x_grid, nx)
+    x_grid = jnp.tile(x_grid[..., jnp.newaxis], (1, x.shape[0]))
+    px = jsp.stats.norm.pdf(x_grid, loc=x, scale=kernel_variance)
+    px = jnp.sum(px, axis=1)
+    px = px / jnp.sum(px) / delta
+    return px
+
+
+def novel_density(
+        rng_key: jnp.DeviceArray, y: jnp.DeviceArray, x_grid: jnp.DeviceArray,
+        x0_pdf: Callable, v_rvs: Callable, e_rvs: Callable,
+        f: Callable, h: Callable, num_samples: int,
+        max_iter: int, kernel_variance: float):
+    num_grid_points = x_grid.shape[0]
+    delta = x_grid[1] - x_grid[0]
+
+    p_filter = jnp.zeros(shape=(max_iter + 1, num_grid_points))
+    p_filter = p_filter.at[0, :].set(x0_pdf(x_grid))
+    p_filter = p_filter.at[0, :].set(p_filter[0, :] / jnp.sum(p_filter[0, :]))
+    p_pred = jnp.ones(shape=(max_iter + 1, num_grid_points)) * jnp.inf
+    rng_keys = random.split(rng_key, num=3)
+    v = v_rvs(rng_keys[0], shape=(max_iter + 1, num_samples))
+    e = e_rvs(rng_keys[1], shape=(max_iter + 1, num_samples))
+
+    for k in range(1, max_iter + 1):
+        x = inversion_sampling(
+            rng_keys[2],
+            x_grid,
+            p_filter[k - 1, :],
+            num_samples=num_samples,
+        )
```

Comment:
Have you tried [`jax.lax.scan`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html) on this code?